interfaces/prompting/constraints.go

```diff
@@ -0,0 +1,263 @@
+	}
+	return abstractPerms, nil
+}
+
+// AbstractPermissionsFromList validates the given permissions list for the
+// given interface and returns a list containing those permissions in the order
+// in which they occur in the list of available permissions for that interface.
+func AbstractPermissionsFromList(iface string, permissions []string) ([]string, error) {
+	availablePerms, ok := interfacePermissionsAvailable[iface]
+	if !ok {
+		return nil, fmt.Errorf("unsupported interface: %q", iface)
+	}
+	permsSet := make(map[string]bool, len(permissions))
+	for _, perm := range permissions {
+		if !strutil.ListContains(availablePerms, perm) {
+			return nil, fmt.Errorf("unsupported permission for %q interface: %q", iface, perm)
+		}
+		permsSet[perm] = true
+	}
+	if len(permsSet) == 0 {
+		return nil, ErrPermissionsListEmpty
+	}
+	permissionsList := make([]string, 0, len(permsSet))
+	for _, perm := range availablePerms {
+		if exists := permsSet[perm]; exists {
+			permissionsList = append(permissionsList, perm)
+		}
+	}
+	return permissionsList, nil
+}
+
+// AbstractPermissionsToAppArmorPermissions returns AppArmor permissions
+// corresponding to the given permissions for the given interface.
+func AbstractPermissionsToAppArmorPermissions(iface string, permissions []string) (interface{}, error) {
+	return abstractPermissionsToAppArmorFilePermissions(iface, permissions)
+}
+
+// AbstractPermissionsToAppArmorFilePermissions returns AppArmor file
+// permissions corresponding to the given permissions for the given interface.
+func abstractPermissionsToAppArmorFilePermissions(iface string, permissions []string) (notify.FilePermission, error) {
+	if len(permissions) == 0 {
+		return notify.FilePermission(0), ErrPermissionsListEmpty
+	}
+	filePermsMap, exists := interfaceFilePermissionsMaps[iface]
+	if !exists {
+		// This should never happen, since iface is checked in the calling function
+		return notify.FilePermission(0), fmt.Errorf("internal error: no AppArmor file permissions map defined for interface: %q", iface)
+	}
+	filePerms := notify.FilePermission(0)
+	for _, perm := range permissions {
+		permMask, exists := filePermsMap[perm]
+		if !exists {
+			// Should not occur, since stored permissions list should have been validated
+			return notify.FilePermission(0), fmt.Errorf("no AppArmor file permission mapping for %q interface with abstract permission: %q", iface, perm)
+		}
+		filePerms |= permMask
+	}
+	if filePerms&(notify.AA_MAY_EXEC|notify.AA_MAY_WRITE|notify.AA_MAY_READ|notify.AA_MAY_APPEND|notify.AA_MAY_CREATE) != 0 {
+		filePerms |= notify.AA_MAY_OPEN
+	}
+	return filePerms, nil
+}
+
+// ValidateConstraintsOutcomeLifespanExpiration returns an error if the given
+// constraints, outcome, lifespan, or duration are invalid, else returns nil.
+func ValidateConstraintsOutcomeLifespanExpiration(iface string, constraints *Constraints, outcome OutcomeType, lifespan LifespanType, expiration *time.Time, currTime time.Time) error {
```

Comment:
this is strange, do we need a struct with the thing passed to this all together and this is a method on it?

Reply:
I agree, this should be a method on the `RequestRule` type, likely `rule.Validate()`, and this is only used when looking over existing rules and making sure they are valid and consistent. Thanks!